streams/src/main/java/org/apache/kafka/streams/processor/internals/SynchronizedPartitionGroup.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import java.util.Set;
+import java.util.function.Function;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.TopicPartition;
+
+class SynchronizedPartitionGroup extends AbstractPartitionGroup {
+
+    private AbstractPartitionGroup wrapped;
+
+    public SynchronizedPartitionGroup(final AbstractPartitionGroup wrapped) {
+        this.wrapped = wrapped;
+    }
+
+    @Override
+    synchronized boolean readyToProcess(final long wallClockTime) {
```

Comment:
I think we need to synchronize all heavily-called methods here. Using partial synchronization wouldn't benefit us much, and it would be confusing IMHO to have unsynchronized methods in `SynchronizedPartitionGroup`.